11_sex_prediction/nbconverted/lasso_range.py

```diff
@@ -0,0 +1,265 @@
+        col_wrap=6, height=4.5, aspect=1.35
+    )
+    g.set_titles('Holdout cancer type: {col_name}')
+    g.set_xlabels('Lasso parameter \n (higher = less regularization)')
+    g.set_ylabels(f'{metric.upper()}')
+    g.set_xticklabels(rotation=70)
+    plt.suptitle(f'LASSO parameter vs. {metric.upper()}, sex prediction', y=1.02)
+
+if output_plots:
+    output_plots_dir.mkdir(exist_ok=True)
+    plt.savefig(output_plots_dir / f'msi_lasso_boxes.png',
+                dpi=200, bbox_inches='tight')
+
+
+# In[7]:
+
+
+# try with a float-valued x-axis
+# this is probably more "correct" than treating each lasso parameter as a
+# category (above plot); here the spaces between parameters reflect their
+# actual real-valued distance in log-space
+# sns.set({'figure.figsize': (12, 5)})
+sns.set_style('ticks')
+
+plot_df = (
+    perf_df[(perf_df.signal == 'signal')]
+      .sort_values(by=['holdout_cancer_type', 'lasso_param'])
+      .reset_index(drop=True)
+)
+plot_df.lasso_param = plot_df.lasso_param.astype(float)
+
+with sns.plotting_context('notebook', font_scale=1.6):
+    g = sns.relplot(
+        data=plot_df,
+        x='lasso_param', y=metric, hue='data_type',
+        hue_order=['train', 'cv', 'test'],
+        kind='line', col='holdout_cancer_type',
+        col_wrap=6, height=4, aspect=1.2
+    )
+    g.set(xscale='log', xlim=(min(plot_df.lasso_param), max(plot_df.lasso_param)))
+    g.set_titles('Holdout cancer type: {col_name}')
+    g.set_xlabels('Lasso parameter \n (higher = less regularization)')
+    g.set_ylabels(f'{metric.upper()}')
+    plt.suptitle(f'LASSO parameter vs. {metric.upper()}, sex prediction', y=1.01)
+
+if output_plots:
+    plt.savefig(output_plots_dir / f'sex_lasso_facets.png',
+                dpi=200, bbox_inches='tight')
+
+
+# ### Plot performance vs. nonzero coefficient count
+# 
+# Since different models (across seeds/folds) have slightly different numbers of nonzero coefficients, the easiest way to do this is to "bin" the models by coefficient count. Here we'll just divide the space of possible feature counts into 10 quantiles, then bin models into those.
+
+# In[8]:
+
+
+plot_df = (
+    perf_df[(perf_df.signal == 'signal')]
+      .merge(nz_coefs_df, left_on=['holdout_cancer_type', 'lasso_param', 'seed', 'fold'],
+             right_on=['cancer_type', 'lasso_param', 'seed', 'fold'])
+      .drop(columns=['cancer_type'])
+      .sort_values(by=['holdout_cancer_type', 'lasso_param'])
+      .reset_index(drop=True)
+)
+plot_df.lasso_param = plot_df.lasso_param.astype(float)
+plot_df.head()
+
+
+# In[9]:
+
+
+sns.set({'figure.figsize': (15, 6)})
+sns.set_style('whitegrid')
+sns.histplot(plot_df.nz_coefs)
+
+quantiles_df = []
+
+for q in np.linspace(0.1, 0.9, 9):
+    quantiles_df.append([q, plot_df.nz_coefs.quantile(q)])
+    plt.gca().axvline(x=plot_df.nz_coefs.quantile(q), color='black', linestyle='--')
+    
+quantiles_df = pd.DataFrame(quantiles_df, columns=['quantile', 'value'])
```

Comment:
This is a pretty rudimentary question, but what do you make of the extreme bimodal distribution of nonzero coefficients? Is the range of lasso parameters you're giving causing them to have almost all or almost no nonzero coefficients? Does that mean it might be too extreme of a range?

Reply:
I'm also concerned if quantiles is the right approach here. The difference between quantile 3 and 4 is  ~8000 genes, whereas the difference between quantile 7 and 8 and ~40. I feel like using the quantiles as the scale for your boxplot is misrepresenting things. Feel free to push back though.

Reply:
These are both great questions. I think since we're still somewhat in the "throw things at the wall and see what sticks" stage of this project I'm not too concerned quite yet with getting these details exactly right. For instance, I'm not totally sure if we'll go forward with the quantile plots or some other way of summarizing the results, so I'll think this through more if the "bins" are the type of visualization we end up preferring.

To answer the first question in slightly more detail, usually when I'm first doing parameter tuning I prefer a slightly larger range, and I think for this project in particular we want to make sure to include really simple models (almost no nonzero coefficients) and really complex un-regularized models (almost all nonzero coefficients) so I chose a range that would cover those cases. But I do think for the next analysis I do (coming soon!) it will be more important to make sure the "middle range" is more densely covered so that's a great point that I agree with.

I'll create an issue for both of these questions, since I do want to keep them in mind going forward.